src/System.Numerics.Tensors/System/Numerics/Tensor.cs

```diff
@@ -1039,6 +1029,117 @@ void IList.RemoveAt(int index)
         }
         #endregion
 
+        #region IEnumerable<T> members
+        public IEnumerator<T> GetEnumerator()
+        {
+            for (int i = 0; i < Length; i++)
+            {
+                yield return GetValue(i);
+            }
+        }
+        #endregion
+
+        #region ICollection<T> members
+        int ICollection<T>.Count => (int)Length;
+
+        void ICollection<T>.Add(T item)
+        {
+            throw new InvalidOperationException();
```

Comment:
It makes me wonder why Tensor even implements the collection interfaces. I know it's not part of the PR, and more of a question to @ericstj

Reply:
@ericstj is on leave for the next 3 months and I'm taking this work over.

I would think this behavior satisfies the collection interfaces because Tensor always returns `IsFixedSize = true`.  Thus you can't add/insert/remove.

From a general perspective, it makes sense to me for these interfaces to be implemented.  Tensor is a collection of values.  It just happens to be n-dimensional.

Reply:
But what are the scenarios for Tensors being operated on as IList or IList\<T\>?

Reply:
I believe the original intention was to match the API surface of Array. See the issue I'm fixing https://github.com/dotnet/corefxlab/issues/1876.

> Also look at what's exposed by T[,] and make sure we have everything.
Originally I just looked at Array so I was missing all the generic interfaces.

For `IEnumerable<T>`, a concrete scenario is with the existing CNTK library.  When you call their evaluation library, they take an `IEnumerable<T>` for the data.  See [CreateBatch](https://github.com/Microsoft/CNTK/blob/73cae11ed6707c5bfe0adc7a2f238da79ab52342/bindings/csharp/CNTKLibraryManagedDll/ShimApiClasses/ValueShim.cs#L104). 

Of course we want to make that exact API better in the future, by not copying the data to a new array, and directly passing the Tensor's memory into their native code.  But my thinking is that there are probably other libraries that take IList/Collection/etc for their data, and implementing these interfaces on Tensor allow for it to work with those libraries.

Reply:
But what's about `IReadOnlyList<T>`? `IList<T>` doesn't inherit from it because of [that issue](https://social.msdn.microsoft.com/Forums/vstudio/en-US/b4fb991a-3f5c-4923-93d4-7cd5c004f859/new-interfaces-ireadonlylist-and-ireadonlydictionary?forum=netfxbcl), so classes should implement it themselves.

> **Why did we not change the existing interfaces to extend the read-only interfaces?**
> It looks like a reasonable assumption that it works because the read-only interfaces are purely a subset of the read-write interfaces. Unfortunately, it is incompatible because at the metadata level every method on every interface has its own slot (which makes explicit interface implementations work).

Reply:
Will add `IReadOnlyList<T>`.